### ISSUE TITLE
Character encoding and European characters

### ISSUE THREAD

**Issue by burkmarr** [Indicia-Team/warehouse#343]
I'm working on uploading some locations, representing transects & sections, into Warehouse1 for eBMS. The Spanish names of the transects/sections often use 'special' characters, e.g:

PUERTOS DE AGÜERIA

When I import these (using Shapefile import) into the Warehouse, they appear differently on the Warehouse website, e.g. the one above looks like this:

![image](https://user-images.githubusercontent.com/9275178/69881585-64f73000-12c5-11ea-9783-4f3a0fd2bf18.png)

And viewing the record with pgAdmin, it looks like this:

![image](https://user-images.githubusercontent.com/9275178/69881841-2d3cb800-12c6-11ea-912f-d0e21b06d5be.png)

(If I copy and paste either of those in here, it looks like this: PUERTOS DE AGÃERIA.)

The name in is correctly displayed in QGIS and the shapefile's character encoding is specified as UTF-8, so I'm pretty sure that it's going into the upload upload process in the correct format. If I run ```SHOW SERVER_ENCODING``` from Warehouse1 pgAdmin the result is 'UTF8', so I would have thought that all should be okay.

I can see plenty of ABLE transects/sections with special characters displayed correctly in the Warehouse. Can anyone throw any light on what might be going on here and what I need to do to ensure that the characters are correctly represented in the Warehouse @johnvanbreda @Gary-van-Breda?

**Comment by johnvanbreda** [comment]
@burkmarr I've noticed this problem before and fortunately had a script to fix the location names after import which you can adapt:
```sql
update locations 
set name=convert_from(convert_to(name, 'latin-1'), 'utf-8'), updated_on=now(), updated_by_id=<your user id>
where <filter>
```

If you have the inclination you could look at the underlying problem, which I suspect centres around line 499 of application/controllers/location.php where the value read from the dbf file is utf8_encoded, which might be unnecessary?

**Comment by burkmarr** [comment]
Thanks @johnvanbreda that script worked on the transects I've already loaded to the Dev Warehouse. I'll leave this open, mark as a bug and assign to myself. Then I will look at the code as time permits.

**Comment by burkmarr** [comment]
@johnvanbreda - I think you are right; that line of code was responsible for messing up my imported characters. I found a workaround by using QGIS to save the shapefile with a different encoding - I used ```Windows 1252```. This preserved the characters in my Shapefile but also withstands the conversion to UTF8 applied by Indicia on import.

I'm not sure if we should remove this or not? What reason did you put it in? Does Indicia require UTF8? If so, then we should perhaps remove this and specify in the documentation that shapefiles should be encoded in UTF8?

On a similar note, I also found that my Location CSV upload incorrectly encoded the same characters. The code responsible for that is line 486 in https://github.com/Indicia-Team/warehouse/blob/master/modules/indicia_svc_import/controllers/services/import.php. However, in this case it actually tests to see if the CSV is UTF8 or not before converting. This confused me because my input file was UTF8 so shouldn't be converted. However, on further investigation I found out there there are two types of UTF8 - UTF8 and UTF8 with BOM. Mine was the type without BOM and the Indicia code is relying on the presence of the BOM to sniff UTF8 and, not finding it, converts it again. Again, my workaround here was to save the file with a different encoding - I chose  ```Windows 1252``` again.

**Comment by johnvanbreda** [comment]
Regarding the shapefile import, we replaced the PHP dbase extension with a different native PHP library a year or tow back, since the PHP dbase extension is not compiled into many PHP distributions making installation tricky. I suspect at this point we didn't test how it handled extended characters and missed the fact that the utf8_encode was no longer required as it was doubling the encoding being done by the library.

For the CSV import, our code is imperfect but was designed to work with the idiosyncrasies of Excel which adds the BOM for CSV data even though non-standard. I recall having some problems with mb_detect_encoding(), either not being reliable or the mbstrings extension not always being available. Having just Googled, there is a potential alternative approach in https://stackoverflow.com/questions/4407854/how-do-i-detect-if-have-to-apply-utf-8-decode-or-encode-on-a-string which might be worth a look.

**Comment by burkmarr** [comment]
Thanks @johnvanbreda.

**Comment by JimBacon** [comment]
Times have moved on but this is still a pain in the neck. I've just been trying to import a species list which contains lots of accented characters in the authority names.

I have the file in an Excel spreadsheet where all the accented characters are fine. These days I can import as xlsx file directly but we have a bug which double encodes it. If I save the spreadsheet as CSV it uses CP1252 encoding which does not support all characters. Even when I got my file in to a UTF8-encoded CSV file, by several manipulations in Notepad++, I was still caught out by omitting the BOM.

Enter a solution. For reading Excel files we use PhpOffice\PhpSpreadsheet\Reader. This can also read CSV files and can determine the file encoding. 

> You may also let PhpSpreadsheet attempt to guess the input encoding. It will do so based on a test for BOM (UTF-8, UTF-16BE, UTF-16LE, UTF-32BE, or UTF-32LE), or by doing heuristic tests for those encodings, falling back to a specifiable encoding (default is CP1252) if all of those tests fail.

Using it relieves us of any need to deal with character encoding ourselves.

I have tested it successfully on a CSV file with encoding of UTF-8 + BOM, UTF-8 without BOM and CP1252. It also simplifies the import code to use PhpOffice for all file-types rather than having a separate method for CSV files. It may introduce some processing overhead but my impression is that's insignificant compared to the database update time. In any event, once the Excel import works, it would be the preferred format to use.
